secboot/secboot_tpm.go

```diff
@@ -318,35 +318,107 @@ var fdeRevealKeyRuntimeMax = "2m"
+	for _, stream := range []string{"stdin", "stdout", "stderr"} {
+		streamFile := filepath.Join(runDir, "fde-reveal-key."+stream)
+		// we want to make sure that the file permissions for stdout are always
+		// 0600, so to ensure this is the case and be robust against bugs, we
+		// always delete the file and re-create it with 0600
+
+		// note that if the file already exists, WriteFile will not change the
+		// permissions, so deleting first is the right thing to do
+		os.Remove(streamFile)
+		if stream == "stdin" {
+			err = ioutil.WriteFile(streamFile, stdin, 0600)
+		} else {
+			err = ioutil.WriteFile(streamFile, nil, 0600)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("cannot create %s for fde-reveal-key: %v", stream, err)
+		}
+	}
+
 	// TODO: put this into a new "systemd/run" package
 	cmd := exec.Command(
 		"systemd-run",
-		"--pipe", "--same-dir", "--wait", "--collect",
+		"--collect",
 		"--service-type=exec",
 		"--quiet",
 		// ensure we get some result from the hook within a
 		// reasonable timeout and output from systemd if
 		// things go wrong
 		fmt.Sprintf("--property=RuntimeMaxSec=%s", fdeRevealKeyRuntimeMax),
-		// this ensures we get useful output for e.g. segfaults
-		`--property=ExecStopPost=/bin/sh -c 'if [ "$EXIT_STATUS" != 0 ]; then echo "service result: $SERVICE_RESULT" 1>&2; fi'`,
 		// Do not allow mounting, this ensures hooks in initrd
 		// can not mess around with ubuntu-data.
 		//
 		// Note that this is not about perfect confinement, more about
 		// making sure that people using the hook know that we do not
 		// want them to mess around outside of just providing unseal.
 		"--property=SystemCallFilter=~@mount",
+		// WORKAROUNDS
+		// workaround the lack of "--pipe"
+		fmt.Sprintf("--property=StandardInput=file:%s/fde-reveal-key.stdin", runDir),
+		// NOTE: these files are manually created above with 0600 because by
+		// default systemd will create them 0644 and we want to be paranoid here
+		fmt.Sprintf("--property=StandardOutput=file:%s/fde-reveal-key.stdout", runDir),
+		fmt.Sprintf("--property=StandardError=file:%s/fde-reveal-key.stderr", runDir),
+		// this ensures we get useful output for e.g. segfaults
+		fmt.Sprintf(`--property=ExecStopPost=/bin/sh -c 'if [ "$EXIT_STATUS" = 0 ]; then touch %[1]s/fde-reveal-key.success; else echo "service result: $SERVICE_RESULT" >%[1]s/fde-reveal-key.failed; fi'`, runDir),
 	)
 	if fdeRevealKeyCommandExtra != nil {
 		cmd.Args = append(cmd.Args, fdeRevealKeyCommandExtra...)
 	}
 	// fde-reveal-key is what we actually need to run
 	cmd.Args = append(cmd.Args, "fde-reveal-key")
-	return cmd
+
+	// ensure we cleanup our tmp files
+	defer func() {
+		if err := os.RemoveAll(runDir); err != nil {
+			logger.Noticef("cannot remove tmp dir: %v", err)
+		}
+	}()
+
+	// run the command
+	output, err = cmd.CombinedOutput()
+	if err != nil {
+		return output, err
+	}
+
+	// this will terminate eventually because of
+	// fdeRevealKeyRuntimeMax above
```

Comment:
maybe we should be paranoid though and go through this loop almost:    

RuntimeMax/10*13/50ms  times ?  (allowing a 30% overhead)

and error then at end of the function?

Reply:
I will do a followup with that